datafusion/common/src/config.rs

```diff
@@ -1612,42 +1623,241 @@ impl TableOptions {
+    /// based on the current format selected. If no format is selected, it returns an error.
     ///
     /// # Parameters
     ///
-    /// * `settings`: A hash map where each key-value pair represents a configuration setting.
+    /// * `key`: The configuration key specifying which setting to adjust, prefixed with the format (e.g., "format.delimiter")
+    ///   for CSV format.
+    /// * `value`: The value to set for the specified configuration key.
     ///
     /// # Returns
     ///
-    /// A result containing the new `TableOptions` instance or an error if any setting could not be applied.
-    pub fn from_string_hash_map(settings: &HashMap<String, String>) -> Result<Self> {
-        let mut ret = Self::default();
-        for (k, v) in settings {
-            ret.set(k, v)?;
+    /// A result indicating success or an error if the key is not recognized, if a format is not specified,
+    /// or if setting the configuration value fails for the specific format.
+    fn set(&mut self, key: &str, value: &str) -> Result<()> {
+        // Extensions are handled in the public `ConfigOptions::set`
+        let (key, rem) = key.split_once('.').unwrap_or((key, ""));
+        match key {
+            "format" => match self {
+                #[cfg(feature = "parquet")]
+                Self::Parquet { options, .. } => options.set(rem, value),
+                Self::Csv { options, .. } => options.set(rem, value),
+                Self::Json { options, .. } => options.set(rem, value),
+            },
+            _ => _config_err!("Config value \"{key}\" not found on TableOptions"),
         }
+    }
+}
 
-        Ok(ret)
+impl TableFormatOptions {
+    /// Constructs a new instance of `TableOptions` with JSON file type and default settings.
+    ///
+    /// # Returns
+    ///
+    /// A new `TableOptions` instance with default configuration values.
+    pub fn new_json() -> Self {
+        Self::Json {
+            options: Default::default(),
+            extensions: Default::default(),
+        }
     }
 
-    /// Modifies the current `TableOptions` instance with settings from a hash map.
+    /// Constructs a new instance of `TableOptions` with CSV file type and default settings.
+    ///
+    /// # Returns
+    ///
+    /// A new `TableOptions` instance with default configuration values.
+    pub fn new_csv() -> Self {
+        Self::Csv {
+            options: Default::default(),
+            extensions: Default::default(),
+        }
+    }
+
+    /// Constructs a new instance of `TableOptions` with Parquet file type and default settings.
+    ///
+    /// # Returns
+    ///
+    /// A new `TableOptions` instance with default configuration values.
+    #[cfg(feature = "parquet")]
+    pub fn new_parquet() -> Self {
+        Self::Parquet {
+            options: Default::default(),
+            extensions: Default::default(),
+        }
+    }
+
+    pub fn csv_options_or_default(&self) -> CsvOptions {
```

Comment:
Likewise, these new public methods should have documentation

Reply:
After reading this some more, I don't undertand the usecase for `csv_options_or_default` (and the other similar methods)

Since this is now an enum, the calling code should directly handle the format in the TableFormatOptions -- this code will silently drop any config options if it isn't CSV which seems like it would potentially mask a bug in the caller

I would expect calling code to do something like

```rust
match TableFormatOptions { 
  Self::Csv => {}
  Self::Json => {}
  Self::Parquet => {{
}
```